src/signature.cpp

```diff
@@ -26,102 +26,51 @@ namespace mococrw {
-            throw MoCOCrWException("Context is empty");
-        }
+        auto keyCtx = _EVP_PKEY_CTX_new(ctx.getKey().internal());
         _EVP_PKEY_sign_init(keyCtx.get());
-        setUpContext(keyCtx.get(), padding);
+
+        setupOpenSSLCtx(ctx, keyCtx, RSAPadding::OperationTypes::Sign);
 
         // This determines the buffer length
         _EVP_PKEY_sign(keyCtx.get(),
                        nullptr,
-                       &siglen,
+                       &sigLen,
                        reinterpret_cast<const unsigned char *>(messageDigest.data()),
                        messageDigest.size());
 
-        _CRYPTO_malloc_init();
-        sig.reset(static_cast<unsigned char *>(_OPENSSL_malloc(siglen)));
+        signatureData.reset(static_cast<unsigned char *>(_OPENSSL_malloc(sigLen)));
         _EVP_PKEY_sign(keyCtx.get(),
-                       sig.get(),
-                       &siglen,
+                       signatureData.get(),
+                       &sigLen,
                        reinterpret_cast<const unsigned char *>(messageDigest.data()),
                        messageDigest.size());
-    }
-    catch (const OpenSSLException &e) {
-        throw MoCOCrWException(e.what());
-    }
-
-    return std::vector<uint8_t>(sig.get(), sig.get() + siglen);
-}
-
-void SignatureUtils::verify(AsymmetricPublicKey &publicKey,
-                            const RSAPadding &padding,
-                            const std::vector<uint8_t> signature,
-                            const std::string message)
-{
-    std::vector<uint8_t> messageDigest;
-
-    if(publicKey.getType() != AsymmetricKey::KeyTypes::RSA){
-        throw MoCOCrWException("Functionality is only supported for RSA keys");
-    }
 
-    try {
-        messageDigest = digestMessage(message, getHashing(padding));
-        verify(publicKey, padding, signature, messageDigest);
     }
     catch (const OpenSSLException &e) {
         throw MoCOCrWException(e.what());
     }
+
+    return std::vector<uint8_t>(signatureData.get(), signatureData.get() + sigLen);
 }
 
-void SignatureUtils::verify(AsymmetricPublicKey &publicKey,
-                            const RSAPadding &padding,
-                            const std::vector<uint8_t> signature,
-                            const std::vector<uint8_t> messageDigest)
+void SignatureUtils::verify(AsymmetricVerifyCtx &ctx,
+                            const std::vector<uint8_t> &signature,
+                            const std::vector<uint8_t> &messageDigest)
 {
-    const auto keyCtx = _EVP_PKEY_CTX_new(publicKey.internal());
-
-    if(publicKey.getType() != AsymmetricKey::KeyTypes::RSA){
-        throw MoCOCrWException("Functionality is only supported for RSA keys");
-    }
-
     try {
+        auto keyCtx = _EVP_PKEY_CTX_new(ctx.getKey().internal());
         _EVP_PKEY_verify_init(keyCtx.get());
-        setUpContext(keyCtx.get(), padding);
+
+        setupOpenSSLCtx(ctx, keyCtx, RSAPadding::OperationTypes::Verify);
```

Comment:
Operation type could be more global? Does not seem to be specific to RSAPadding.